### PR TITLE
Future3/plugin --> Future3/webapp

### PR DESCRIPTION
**Pull request by ChisSoc**
Merge current status of Future3/plugin into Future3/webapp

As discussed: this will break a few things regarding the RPC Format

We new have package.plugin(.method) as call paramters. Return value in response is only set if, 'id' is set. Response does not included the call signature, but only the id and the result!

Call signature for changing volume over RPC has changed as well:
~~~
package=volume, plugin=ctrl, method= (as before: mute/unmute/set/get)
~~~

Note:
- MPD player not yet adapted! So this will not do anything beyond playing the startup sound!
- CLI not adapted
- Webapp not adapted

### CODE REVIEW

**Comment by ChisSoc** [comment]
The PlayerMPD is ported. New file location: components/playermpd

Note: The is an not yet solved issue: The MPD player does not store its status to disc at the moment. That has to do with the de-coupling of JunkeBox and player plugin. That will be done soon, but needs some additional work. 

The is now quite a few deprecated files. This I can remove if you agree. I kept them at the moment for reference.

File: src/jukebox/jukebox/plugs.py [hunk 1]
```diff
@@ -0,0 +1,660 @@
...
+    Also note that Exceptions in the Thread must be handled in the Thread and are not propagated to the main Thread.
+    All threads are started as daemon threads with terminate upon main program termination.
+    There is not stop-thread mechanism. This is intended for short lived threads.
+    :param args: Arguments passed to callable
+    :param kwargs: Keyword arguments passed to callable
+    :return: The return value from the called function, or, if started as thread the thread object
+    """
+    if logger_call.isEnabledFor(logging.DEBUG):
+        m = f".{method}" if method is not None else ''
+        logger_call.debug(f"Calling: {package}.{plugin}{m}(args={args}, kwargs={kwargs})")
```
Comment (pabera) [inline src/jukebox/jukebox/plugs.py:572]: I was able to run the PR in the Docker environment successfully. this command though was called once a second. Is this intentional?

```
jukebox    | 20.07.2021 12:29:05 -  572:plugs.py             - jb.plugin.call - Thread-238 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:06 -  572:plugs.py             - jb.plugin.call - Thread-239 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:06 -  572:plugs.py             - jb.plugin.call - Thread-240 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:06 -  572:plugs.py             - jb.plugin.call - Thread-241 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:06 -  572:plugs.py             - jb.plugin.call - Thread-242 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:07 -  572:plugs.py             - jb.plugin.call - Thread-243 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:07 -  572:plugs.py             - jb.plugin.call - Thread-244 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:07 -  572:plugs.py             - jb.plugin.call - Thread-245 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:07 -  572:plugs.py             - jb.plugin.call - Thread-246 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:08 -  572:plugs.py             - jb.plugin.call - Thread-247 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:08 -  572:plugs.py             - jb.plugin.call - Thread-248 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:08 -  572:plugs.py             - jb.plugin.call - Thread-249 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:08 -  572:plugs.py             - jb.plugin.call - Thread-250 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:09 -  572:plugs.py             - jb.plugin.call - Thread-251 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:09 -  572:plugs.py             - jb.plugin.call - Thread-252 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:09 -  572:plugs.py             - jb.plugin.call - Thread-253 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:09 -  572:plugs.py             - jb.plugin.call - Thread-254 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:10 -  572:plugs.py             - jb.plugin.call - Thread-255 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
jukebox    | 20.07.2021 12:29:10 -  572:plugs.py             - jb.plugin.call - Thread-256 - DEBUG    - Calling: volume.ctrl.get_volume(args=(), kwargs=None)
```

File: resources/default-settings/jukebox.default.yaml [hunk 2]
```diff
@@ -1,25 +1,33 @@
...
-    playlists_path: ../../shared/playlists
+    # call_parameters: [-d 2]
+    service: auto
 player:
     status_file: ../../shared/settings/music_player_status.json
 modules:
-    volume: jukebox.alsaif
-    player: jukebox.mpd
-    others: [rfid]
+    named:
```
[the remark this answers is not in the document]
Comment (ChisSoc) [reply resources/default-settings/jukebox.default.yaml:6]: This is the plugins to load. Naming is preliminary and I am not fixed on it. Suggestions? I tired to make it a little clearer. Maybe 'plugins' instead if 'modules'?

File: resources/default-settings/jukebox.default.yaml [hunk 3]
```diff
@@ -1,25 +1,33 @@
...
+        volume: volume # Load this first
+        jingle: jingle
+        player: playermpd
```
[the remark this answers is not in the document]
Comment (ChisSoc) [reply resources/default-settings/jukebox.default.yaml:9]: Well, we need a player plugin that dynamically selects which actual player (mpd, spotify) to use. As we currently only have mpd it is loaded as 'player'. The interface 'playermpd' exposes is not the final one of the 'generic player' you have in mind.

File: src/jukebox/components/alsaif/__init__.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            self._instance = AlsaCtrl()
+        return self._instance
+
+
+# ---------------------------------------------------------------------------
+# Jingle Service for WAVE by ALSA
+# ---------------------------------------------------------------------------
+
+@plugin.register
+class AlsaWave:
```
[the remark this answers is not in the document]
Comment (ChisSoc) [reply src/jukebox/components/alsaif/__init__.py]: Part1) Fully agree. One file per plugin. It is changed - it was a coding artefact from quickly checking out the concept with the Factory Pattern. 
Part 2) There is a misunderstanding: AlsaWave is not registered as jingle plugin. It registers itself as player service with the JingleFactory of the jingle plugin. Complicated, I know 
Part 3) Using a wave startup sound (playing directly from python with the code from Arne): 1.1s startup time. Playing the same from MP3: 4.55s startup time. With the current code everybody can choose (and it is done automatically based on file ending)

File: src/jukebox/components/jingle/__init__.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    play(cfg['jingle']['shutdown_sound'])
+
+
+# ---------------------------------------------------------------------------
+# MP3 Jingle Service
+# ---------------------------------------------------------------------------
+# A generic mp3 player service for the jingle playback
+
+@plugin.register(auto_tag=True)
+class JingleMp3Play:
```
[the remark this answers is not in the document]
Comment (ChisSoc) [reply src/jukebox/components/jingle/__init__.py]: Done. I want it very clear, that this is only for jingle playback and not for normal MP3 playback (which would be done through MPD). So I called it 'jinglemp3'

File: src/jukebox/jukebox/plugs.py (hunk 1 above)
Comment (ChisSoc) [reply src/jukebox/jukebox/plugs.py:610]: Expected, as the playermpd checks the volume though ALSA as part of the status update. Not very use, I agree. Changed.

**Comment by ChisSoc** [comment]
A lot of clean up :-) 
And proper shutdown handling of the plugins.

A very useful thing for UI developers might be this: During startup, a text file gets written out with all the loaded plugins including all the Callables and the python documentation for those Callables (if it exists).

Output is here: shared/settings/plugin_reference.txt (but can be configured in jukebox.yaml)

File: resources/default-settings/jukebox.default.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (pabera) [reply resources/default-settings/jukebox.default.yaml:6]: I was referring to `named` .. I am clear on `modules` of course ;-)

File: src/jukebox/components/alsaif/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pabera) [reply src/jukebox/components/alsaif/__init__.py]: 1.1s vs 4.5s is a big difference. I hadn't had that on my mind at all. all good here then

File: src/jukebox/components/jingle/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pabera) [reply src/jukebox/components/jingle/__init__.py]: makes sense now!

**Comment by pabera** [comment]
I am to merge this!

**Review by pabera — APPROVED** [review]

File: src/jukebox/components/playermpd/__init__.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        # Need to make sure we are not in the process of a status poll
+        # This currently causes 250 ms delay in shutdown (which otherwise takes 6 ms)
+        # --> Change it (self.status_thread.cancel() ?)
+        time.sleep(self.mpd_status_poll_interval)
+        # self.status_thread.cancel()
+        self.mpd_client.disconnect()
+        self.nvm.save_all()
+
+    def connect(self):
+        self.mpd_client.connect(self.mpd_host, 6600)
```
Comment (pabera) [inline src/jukebox/components/playermpd/__init__.py:65]: I tried to run this in my Docker env and I received this error. Could `self.mpd_host` be empty for some reason?

```
jukebox    | 22.07.2021 15:47:29 -  514:plugs.py        - jb.plugin            - MainThread      - ERROR    - Failed to load package: components.playermpd
jukebox    | 22.07.2021 15:47:29 -  515:plugs.py        - jb.plugin            - MainThread      - ERROR    - Reason: AttributeError: 'NoneType' object has no attribute 'startswith'
jukebox    | Traceback (most recent call last):
jukebox    |   File "/home/pi/RPi-Jukebox-RFID/src/jukebox/run_jukebox.py", line 56, in <module>
jukebox    |     main()
jukebox    |   File "/home/pi/RPi-Jukebox-RFID/src/jukebox/run_jukebox.py", line 52, in main
jukebox    |     myjukebox.run()
jukebox    |   File "/home/pi/RPi-Jukebox-RFID/src/jukebox/jukebox/daemon.py", line 99, in run
jukebox    |     plugin.load_all_named(plugins_named, prefix='components')
jukebox    |   File "/home/pi/RPi-Jukebox-RFID/src/jukebox/jukebox/plugs.py", line 535, in load_all_named
jukebox    |     raise e
jukebox    |   File "/home/pi/RPi-Jukebox-RFID/src/jukebox/jukebox/plugs.py", line 530, in load_all_named
jukebox    |     load(package, load_as, prefix)
jukebox    |   File "/home/pi/RPi-Jukebox-RFID/src/jukebox/jukebox/plugs.py", line 516, in load
jukebox    |     raise e
jukebox    |   File "/home/pi/RPi-Jukebox-RFID/src/jukebox/jukebox/plugs.py", line 508, in load
jukebox    |     _PLUGINS[load_as].module = importlib.import_module(f'{package}', 'pkg')
jukebox    |   File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
jukebox    |     return _bootstrap._gcd_import(name[level:], package, level)
jukebox    |   File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
jukebox    |   File "<frozen importlib._bootstrap>", line 983, in _find_and_load
jukebox    |   File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
jukebox    |   File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
jukebox    |   File "<frozen importlib._bootstrap_external>", line 728, in exec_module
jukebox    |   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
jukebox    |   File "/home/pi/RPi-Jukebox-RFID/src/jukebox/components/playermpd/__init__.py", line 298, in <module>
jukebox    |     player_ctrl = PlayerMPD()
jukebox    |   File "/home/pi/RPi-Jukebox-RFID/src/jukebox/components/playermpd/__init__.py", line 30, in __init__
jukebox    |     self.connect()
jukebox    |   File "/home/pi/RPi-Jukebox-RFID/src/jukebox/components/playermpd/__init__.py", line 65, in connect
jukebox    |     self.mpd_client.connect(self.mpd_host, 6600)
jukebox    |   File "/usr/local/lib/python3.7/dist-packages/mpd/base.py", line 782, in connect
jukebox    |     if host.startswith("/"):
jukebox    | AttributeError: 'NoneType' object has no attribute 'startswith'
```

Comment (pabera) [reply src/jukebox/components/playermpd/__init__.py:65]: Nevermind, I believe I had not updated the `jukebox.yaml`

**Comment by pabera** [comment]
> Output is here: shared/settings/plugin_reference.txt (but can be configured in jukebox.yaml)  

This is really cool! Already helped me to navigate through some of the plugins and functions!

File: src/jukebox/components/playermpd/__init__.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    def playlistinfo(self):
+        playlistinfo = (self.mpd_retry_with_mutex(self.mpd_client.playlistinfo))
+        return playlistinfo
+
+    # Attention: MPD.listal will consume a lot of memory with large libs.. should be refactored at some point
+    @plugs.tag
+    def list_all_dirs(self):
+        result = self.mpd_retry_with_mutex(self.mpd_client.listall)
+        # list = [entry for entry in list if 'directory' in entry]
+        return result
```
Comment (pabera) [inline src/jukebox/components/playermpd/__init__.py:285]: I tried using this in my webapp with this payload

```
{ package: "player", plugin: "ctrl", name: "ctrl", method: "list_all_dirs", kwargs: {} }
```

It works properly, the package and plugin gets called perfectly and the debugger returns the following

```
-   58:server.py       - jb.rpc.server        - MainThread      - DEBUG    - Request: {'package': 'player', 'plugin': 'ctrl', 'name': 'ctrl', 'method': 'list_all_dirs', 'kwargs': {}}
-  610:plugs.py        - jb.plugin.call       - MainThread      - DEBUG    - Calling: player.ctrl.list_all_dirs(args=(), kwargs={})
-  111:server.py       - jb.rpc.server        - MainThread      - DEBUG    - Sending response: {'result': None}
```

You can already see, the return result is `{'result': None}` , but I am actually expecting the list of actual directories. When inspecting `result` using the debugger, it does show the resultset of `listall` but it does not get passed on.

File: src/jukebox/jukebox/plugs.py [hunk 8]
```diff
@@ -0,0 +1,726 @@
...
+    >>> myinst2 = MyClass2()
+    >>> plugin.register(myinst2, name='myinst2')
+
+Naming convention:
+  package      : (a) Either a python package
+              : (b) or a plugin package (which si the python package but probably loaded under a different name inside plugs)
+  plugin       : (a) An object from the package that can be accessed through the plugs call function
+               :     (i.e. a function or a class instance)
+               : (b) the string name to above object
+  name         :     the string name of the plugin object for registration
```
Comment (pabera) [inline src/jukebox/jukebox/plugs.py:82]: Here is a payload I created after this convention

```
{
  package: "player",
  plugin: "ctrl",
  name: "ctrl",
  method: "list_all_dirs",
  kwargs: {}
}
```

Will `plugin` and `name` always be the same? Do you have an example where those values are different

File: src/jukebox/components/playermpd/__init__.py (hunk 7 above)
Comment (pabera) [reply src/jukebox/components/playermpd/__init__.py:285]: Never mind, after reading your code more carefully, I saw the notion of `id` to be a required field. I remember you told us about it. It now works

File: src/jukebox/jukebox/plugs.py (hunk 8 above)
Comment (ChisSoc) [reply src/jukebox/jukebox/plugs.py:82]: A slight of misunderstanding here: Inside plugs.py, I need to differentiate between the python package name (plugin) and the alias name under which is well be accessible through plugs.call() / get() / exists() which may be the same or another (in this case "name"). From the outside there is only three elements: package, plugin, method. Attention: in this case the aliased name is used here. Take a look at plugs.call(). 
Also: When going through the RPC, the call signature is defined by the RPC. Which has the 3 elements package, plugin, method. (name is not used here - plugin is always the aliased name). There is bit of documentation in [rpc/server.py](https://github.com/ChisSoc/RPi-Jukebox-RFID/blob/0d57413fe55e908c824cd3baad890665700f3d58/src/jukebox/jukebox/rpc/server.py#L65):
```
            # {
            #   'package'  : str  # The plugin package loaded from python module
            #   'plugin'   : str  # The plugin object to be accessed from the package (i.e. function or class instance)
            #   'method'   : str  # (optional) The method of the class instance
            #   'args'     : [ ]  # (optional) Positional arguments as list
            #   'kwargs'   : { }  # (optional) Keyword arguments as dictionary
            #   'as_thread': bool # (optional) start call in separate thread
            #   'id'       : Any  # (optional) Round-trip id for response
            #   'tsp'      : Any  # (optional) measure and return total processing time for the call request
            # }

```
If you look at shared/settings/plugin_reference.txt it is
first column: package
second column: plugin
third column: method (if any / optional)

**Pull request merged**